This change unfortunately breaks the ``build-deb`` and ``build-rpm`` build bundles. See the logs below.

The reason why this breaks is because of the way the ``build-deb`` and ``build-rpm`` bundles work. Here is a brief overview.

For each directory under ``contrib/builder/[deb, rpm]/`` we build the dockerfile that is listed there, and start a container for each of these distribution flavors. These containers are run docker in docker on the main docker container. The main Dockerfile, installs containerd and runc's binaries in that container's /usr/local/bin, which isn't available in the child containers where the deb and rpm packages are created.

Since they are not available, we get an error when trying to copy the files.

``cp: cannot stat '/usr/local/bin/runc': No such file or directory``

I'm currently working on a way to get these files added to the packages, hopefully done today, since code freeze is tomorrow.. In the meantime, maybe we can add a check to the copy commands where we don't copy if they don't exist. 


```
Step 7 : RUN dpkg-buildpackage -uc -us
 ---> Running in e0a0bfdfc589
dpkg-buildpackage: source package docker-engine
dpkg-buildpackage: source version 1.11.0~dev~git20160321.140913.0.5008a34-0~jessie
dpkg-buildpackage: source distribution jessie
dpkg-buildpackage: source changed by Docker <support@docker.com>
 dpkg-source --before-build docker
dpkg-buildpackage: host architecture amd64
 debian/rules clean
dh clean --with=bash-completion --with=systemd
   dh_testdir
   dh_auto_clean
   dh_clean
 dpkg-source -b docker
dpkg-source: warning: no source format specified in debian/source/format, see dpkg-source(1)
dpkg-source: warning: source directory 'docker' is not <sourcepackage>-<upstreamversion> 'docker-engine-1.11.0~dev~git20160321.140913.0.5008a34'
dpkg-source: info: using source format `1.0'
dpkg-source: info: building docker-engine in docker-engine_1.11.0~dev~git20160321.140913.0.5008a34-0~jessie.tar.gz
dpkg-source: info: building docker-engine in docker-engine_1.11.0~dev~git20160321.140913.0.5008a34-0~jessie.dsc
 debian/rules build
dh build --with=bash-completion --with=systemd
   dh_testdir
   dh_auto_configure
   debian/rules override_dh_auto_build
make[1]: Entering directory '/usr/src/docker'
./hack/make.sh dynbinary
# WARNING! I don't seem to be running in a Docker container.
# The result of this command might be an incorrect build, and will not be
# officially supported.
#
# Try this instead: make all
#

bundles/1.11.0-dev already exists. Removing.

---> Making bundle: dynbinary (in bundles/1.11.0-dev/dynbinary)
Building: bundles/1.11.0-dev/dynbinary/docker-1.11.0-dev
Created binary: bundles/1.11.0-dev/dynbinary/docker-1.11.0-dev
Copying nested executables into bundles/1.11.0-dev/dynbinary
+++ cp /usr/local/bin/runc bundles/1.11.0-dev/dynbinary/
cp: cannot stat '/usr/local/bin/runc': No such file or directory
make[1]: *** [override_dh_auto_build] Error 1
debian/rules:13: recipe for target 'override_dh_auto_build' failed
make[1]: Leaving directory '/usr/src/docker'
debian/rules:40: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
```

--- a/hack/make/binary
+++ b/hack/make/binary
@@ -62,3 +62,18 @@ echo "Created binary: $DEST/$BINARY_FULLNAME"
 ln -sf "$BINARY_FULLNAME" "$DEST/docker$BINARY_EXTENSION"
 
 hash_files "$DEST/$BINARY_FULLNAME"
+
+# Add nested executables to bundle dir so we have complete set of
+# them available, but only if the native OS/ARCH is the same as the
+# OS/ARCH of the build target
+if [ "$(go env GOOS)/$(go env GOARCH)" == "$(go env GOHOSTOS)/$(go env GOHOSTARCH)" ]; then
+	echo "Copying nested executables into $DEST"
+	(set -x
+	if [ -x /usr/local/bin/runc ]; then
+		cp /usr/local/bin/runc $DEST/
+		cp /usr/local/bin/ctr $DEST/
+		cp /usr/local/bin/containerd $DEST/
+		cp /usr/local/bin/containerd-shim $DEST/
+	fi
+	)
+fi

--- a/hack/make/gccgo
+++ b/hack/make/gccgo
@@ -27,3 +27,18 @@ echo "Created binary: $DEST/$BINARY_FULLNAME"
 ln -sf "$BINARY_FULLNAME" "$DEST/docker$BINARY_EXTENSION"
 
 hash_files "$DEST/$BINARY_FULLNAME"
+
+# Add nested executables to bundle dir so we have complete set of
+# them available, but only if the native OS/ARCH is the same as the
+# OS/ARCH of the build target
+if [ "$(go env GOOS)/$(go env GOARCH)" == "$(go env GOHOSTOS)/$(go env GOHOSTARCH)" ]; then
+    echo "Copying nested executables into $DEST"
+    (set -x
+    if [ -x /usr/local/bin/runc ]; then
+        cp /usr/local/bin/runc $DEST/
+        cp /usr/local/bin/ctr $DEST/
+        cp /usr/local/bin/containerd $DEST/
+        cp /usr/local/bin/containerd-shim $DEST/
+    fi
+    )
+fi

